[JVM] SerializationException: Serializer for class 'kotlin.Any' is not found on JVM platform

# Problem

  The PostHog KMP library fails to initialize on JVM platform with a serialization error when trying to create a serializer for
  Map<String, Any?>.

 ## Error Details

  Caused by: java.lang.ExceptionInInitializerError: Exception kotlinx.serialization.SerializationException: Serializer for class
  'kotlin.Any' is not found.
  Please ensure that class is marked as '@Serializable' and that the serialization compiler plugin is applied.

  Root cause: Line in PostHog.jvm.kt:
  private val mapSerializer = kotlinx.serialization.serializer<Map<String, Any?>>()

 ## Environment

  - Platform: JVM/Desktop
  - PostHog KMP version: 0.1.2
  - Kotlin: 2.3.0
  - kotlinx.serialization: 1.7.1

##  Impact

  - PostHog completely fails to initialize on JVM
  - All analytics events are lost on desktop applications
  - App continues to run but with broken analytics tracking

 ## Reproduction Steps

  1. Add PostHog KMP dependency to a JVM/Desktop Kotlin Multiplatform project
  2. Try to initialize PostHog with PostHog.setup(config, context)
  3. Observe ExceptionInInitializerError during initialization

 ## Root Cause Analysis

  Kotlin serialization doesn't have a built-in serializer for Any? type. The library attempts to create a generic serializer for
  Map<String, Any?> but fails because Any? cannot be automatically serialized without explicit type information.

#  Proposed Solution

  Replace the problematic line with a custom KSerializer<Map<String, Any?>> that handles type-safe serialization:
```
  @OptIn(ExperimentalSerializationApi::class)
  object MapStringAnySerializer : KSerializer<Map<String, Any?>> {
      override val descriptor: SerialDescriptor = buildClassSerialDescriptor("MapStringAny")

      private fun Any?.toJsonElement(): JsonElement = when (this) {
          null -> JsonNull
          is String -> JsonPrimitive(this)
          is Number -> JsonPrimitive(this)
          is Boolean -> JsonPrimitive(this)
          is List<*> -> buildJsonArray {
              this@toJsonElement.forEach { item ->
                  add(item.toJsonElement())
              }
          }
          is Map<*, *> -> buildJsonObject {
              this@toJsonElement.forEach { (key, value) ->
                  put(key.toString(), value.toJsonElement())
              }
          }
          else -> JsonPrimitive(this.toString())
      }

      override fun serialize(encoder: Encoder, value: Map<String, Any?>) {
          val jsonObject = buildJsonObject {
              value.forEach { (key, v) ->
                  put(key, v.toJsonElement())
              }
          }
          encoder.encodeSerializableValue(JsonObject.serializer(), jsonObject)
      }

      override fun deserialize(decoder: Decoder): Map<String, Any?> {
          val jsonObject = decoder.decodeSerializableValue(JsonObject.serializer())
          return jsonObject.mapValues { (_, v) ->
              when (v) {
                  is JsonPrimitive -> when {
                      v.isString -> v.content
                      v.booleanOrNull != null -> v.boolean
                      v.intOrNull != null -> v.int
                      v.longOrNull != null -> v.long
                      v.floatOrNull != null -> v.float
                      v.doubleOrNull != null -> v.double
                      else -> v.content
                  }
                  is JsonNull -> null
                  is JsonArray -> v.map { it.toString() }
                  is JsonObject -> v.toString()
                  else -> v.toString()
              }
          }
      }
  }

  // Replace the problematic line with:
  private val mapSerializer = MapStringAnySerializer
```
##  Additional Notes

  - This issue only affects JVM platform; other platforms (Android, iOS, JS, WASM) work correctly
  - The custom serializer properly handles nested objects and arrays as actual JSON structures (not stringified)
  - Solution maintains backward compatibility and doesn't change the public API

##  Workaround

  Until this is fixed, users can:
  1. Use the custom serializer implementation above
  2. Or catch the initialization error and gracefully degrade analytics functionality

--- a/posthog-kmp/src/jvmMain/kotlin/io/github/samuolis/posthog/PostHog.jvm.kt
+++ b/posthog-kmp/src/jvmMain/kotlin/io/github/samuolis/posthog/PostHog.jvm.kt
@@ -384,7 +384,7 @@ private fun getCurrentISOTimestamp(): String {
 
 // Custom serializer for Map<String, Any?>
 @OptIn(ExperimentalSerializationApi::class)
-object MapStringAnySerializer : KSerializer<Map<String, Any?>> {
+private object MapStringAnySerializer : KSerializer<Map<String, Any?>> {
     override val descriptor: SerialDescriptor = buildClassSerialDescriptor("MapStringAny")
 
     private fun Any?.toJsonElement(): JsonElement = when (this) {

--- a/posthog-kmp/src/jvmMain/kotlin/io/github/samuolis/posthog/PostHog.jvm.kt
+++ b/posthog-kmp/src/jvmMain/kotlin/io/github/samuolis/posthog/PostHog.jvm.kt
@@ -9,6 +9,9 @@ import io.ktor.http.*
 import io.ktor.serialization.kotlinx.json.*
 import kotlinx.coroutines.*
 import kotlinx.serialization.json.*
+import kotlinx.serialization.*
+import kotlinx.serialization.descriptors.*
+import kotlinx.serialization.encoding.*
 import java.time.Instant
 import java.time.format.DateTimeFormatter
 import java.util.*
@@ -92,6 +95,7 @@ internal actual fun platformCapture(event: String, properties: Map<String, Any?>
             scope.launch { flushEvents() }
         }
     }
+    println("PosthogAnalyticsManager platformCapture: Sent event $event $eventProperties")
 }
 
 internal actual fun platformScreen(screenName: String, properties: Map<String, Any?>?) {
@@ -302,14 +306,16 @@ private suspend fun flushEvents() {
         "batch" to batch
     )
 
+    val requestBody = json.encodeToString(mapSerializer, body)
+
     try {
         val response = client.post("${cfg.host}/batch") {
             contentType(ContentType.Application.Json)
-            setBody(json.encodeToString(mapSerializer, body))
+            setBody(requestBody)
         }
 
         if (cfg.debug) {
-            println("[PostHog] Flushed ${eventsToSend.size} events: ${response.status}")
+            println("[PostHog] Flushed ${eventsToSend.size} events: ${response.status}\n${if (!response.status.isSuccess()) requestBody+"\n"+response else ""}")
         }
     } catch (e: Exception) {
         if (cfg.debug) {
@@ -376,5 +382,58 @@ private fun getCurrentISOTimestamp(): String {
     return DateTimeFormatter.ISO_INSTANT.format(Instant.now())
 }
 
-// Simple map serializer for the body
-private val mapSerializer = kotlinx.serialization.serializer<Map<String, Any?>>()
+// Custom serializer for Map<String, Any?>
+@OptIn(ExperimentalSerializationApi::class)
+object MapStringAnySerializer : KSerializer<Map<String, Any?>> {
+    override val descriptor: SerialDescriptor = buildClassSerialDescriptor("MapStringAny")
+
+    private fun Any?.toJsonElement(): JsonElement = when (this) {
+        null -> JsonNull
+        is String -> JsonPrimitive(this)
+        is Number -> JsonPrimitive(this)
+        is Boolean -> JsonPrimitive(this)
+        is List<*> -> buildJsonArray {
+            this@toJsonElement.forEach { item ->
+                add(item.toJsonElement())
+            }
+        }
+        is Map<*, *> -> buildJsonObject {
+            this@toJsonElement.forEach { (key, value) ->
+                put(key.toString(), value.toJsonElement())
+            }
+        }
+        else -> JsonPrimitive(this.toString())
+    }
+
+    override fun serialize(encoder: Encoder, value: Map<String, Any?>) {
+        val jsonObject = buildJsonObject {
+            value.forEach { (key, v) ->
+                put(key, v.toJsonElement())
+            }
+        }
+        encoder.encodeSerializableValue(JsonObject.serializer(), jsonObject)
+    }
+
+    override fun deserialize(decoder: Decoder): Map<String, Any?> {
+        // For PostHog we mainly need serialization, but provide basic deserialization
+        val jsonObject = decoder.decodeSerializableValue(JsonObject.serializer())
+        return jsonObject.mapValues { (_, v) ->
+            when (v) {
+                is JsonPrimitive -> when {
+                    v.isString -> v.content
+                    v.booleanOrNull != null -> v.boolean
+                    v.intOrNull != null -> v.int
+                    v.longOrNull != null -> v.long
+                    v.floatOrNull != null -> v.float
+                    v.doubleOrNull != null -> v.double
+                    else -> v.content
+                }
+                is JsonNull -> null
+                else -> v.toString()
+            }
+        }
+    }
+}
+
+// Custom map serializer for the body
+private val mapSerializer = MapStringAnySerializer

